_posts/2024-05-14-registry-dereference.md

```diff
@@ -0,0 +1,285 @@
+---
+layout: post
+title: "Apicurio Registry - JSON Schema dereference"
+date:   2024-05-14 16:00:00
+author: carles
+categories: registry openapi jsonchema asyncapi avro
+---
+
+As it can be seen in the [documentation](https://www.apicur.io/registry/docs/apicurio-registry/2.5.x/getting-started/assembly-managing-registry-artifacts-api.html), Apicurio Registry allows to manage artifacts with references.
+One of the cool features we've added on top of this is the possibility of using a dereference parameter for certain API endpoints that, when used, allows you to fetch the full content of an artifact with all the references inlined within the same content.
+This is especially useful in certain contexts to reduce the number of HTTP requests, like, as we will see in this blog, in the Kafka Serializers and Deserializers.
+
+---
+
+Intro
+===
+
+The code example used for this blog post can be found in the Apicurio Registry [examples](https://github.com/Apicurio/apicurio-registry/blob/2.6.x/examples/serdes-with-references/src/main/java/io/apicurio/registry/examples/references/JsonSerdeReferencesDereferencedExample.java).
+
+One of the limitations of the JSON Schema format is that, given some data that adheres to a particular schema it does not allow to discover the full schema from that content. That's the main reason why, for this serde, the schema being used has to be registered upfront, and that's what you see
+in the code example from line [57](https://github.com/Apicurio/apicurio-registry/blob/2.6.x/examples/serdes-with-references/src/main/java/io/apicurio/registry/examples/references/JsonSerdeReferencesDereferencedExample.java#L57) to line [132](https://github.com/Apicurio/apicurio-registry/blob/2.6.x/examples/serdes-with-references/src/main/java/io/apicurio/registry/examples/references/JsonSerdeReferencesDereferencedExample.java#L132).
+Where the main schema _city.json_ is registered with all the references in the hierarchy. Along those lines, a full hierarchy of schemas is registered in Apicurio Registry. The hierarchy can be described as follows and represented as a tree:
+
+```
+                          cityzen
+     |          |                     |                   |
+    city   qualification        sample.address     citizenIdentifier
+     |                                                     |
+ qualification                                       qualification
+    
+```
+
+As you may notice, there are three different files with the same name. This is done on purpose to demonstrate that you can have different schemas with the same name within the same hierarchy.
```

Comment:
I didn't know they were file names, should I have known?

Reply:
Not necessarily. They might be URLs to other schemas, they just happen to be files in this example.